test/source/libs/gen_allocation.py

```diff
@@ -121,48 +121,112 @@ def make_funcs(file, category, types):
+        name = t['name'].replace(' ', '_')
+        
+        # Plain type test
+        file.write(subcase.format(t['name'], name, t['res']))
         if 'type' in t:
             file.write('    CHECK(parameters[0].type == "{0}");\n'.format(t['type']))
         file.write('  }')
+        
+        # Single pointer indirection test
+        file.write(subcase.format(t['name']+'*', 'ptr_'+name, '%rdi'))
+        file.write('    CHECK(parameters[0].type == "Pointer64");\n')
+        file.write('  }')
+        
+        # Double pointer indirection test
+        file.write(subcase.format(t['name']+'**', 'ptr_ptr_'+name, '%rdi'))
+        file.write('    CHECK(parameters[0].type == "Pointer64");\n')
+        file.write('  }')
     
     file.write("\n}\n")
 
-# ------------------------------------------ #
+def write_headers(test_file, func_file):
+    func_file.write(
+"""
+// Functions to test register allocation
+#include <complex.h>
+#include <iostream>
+""")
+    test_file.write(
+"""
+// Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+// Spack Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+#include <doctest/doctest.h>
+#include <string.h>
+
+#include "smeagle/smeagle.h"
+
+auto const& get_one(smeagle::Corpus const& corpus, char const* name) {
+  auto const& funcs = corpus.getFunctions();
+
+  return *std::find_if(funcs.begin(), funcs.end(), [name](smeagle::abi_description const& d) {
+    return d.function_name == name;
+  });
+}
+
+namespace {
+  auto corpus = smeagle::Smeagle("liballocation.so").parse();
+}
 
-with open(options.test_filename, 'w') as test_file:
-    with open(options.func_filename, 'w') as func_file:
-        for t in types:
-            make_funcs(func_file, t['category'], t['types'])
-            make_tests(test_file, t['category'], t['types'])
-        
-        # Null type
-        func_file.write(
+"""
+    )    
+
+def write_null_tests(test_file, func_file):
+    func_file.write(
 """
 // Register Allocation - Null Type
 extern "C" void test_void(){}
+extern "C" void test_ptr_void(void* x){}
+extern "C" void test_ptr_ptr_void(void** x){}
 """)
 
-        test_file.write(
+    test_file.write(
 """
 TEST_CASE("Register Allocation - Null Types") {
   SUBCASE("void") {
     auto const& func = get_one(corpus, "test_void");
     auto const& parameters = func.parameters;
     CHECK(parameters.size() == 0UL);
   }
+  SUBCASE("void") {
+    auto const& func = get_one(corpus, "test_ptr_void");
+    auto const& parameters = func.parameters;
+    CHECK(parameters[0].location == "%rdi");
+    CHECK(parameters[0].type == "Pointer64");
+  }
+  SUBCASE("void") {
+    auto const& func = get_one(corpus, "test_ptr_ptr_void");
+    auto const& parameters = func.parameters;
+    CHECK(parameters[0].location == "%rdi");
+    CHECK(parameters[0].type == "Pointer64");
+  }
 }
 """)
+
+# ------------------------------------------ #
```

Comment:
Huh, what's the benefit of writing this at runtime vs. just having the file?

Reply:
I added <10 lines of Python to generate hundreds of lines of C++. When we get reference working, we'll have the same benefit.

Reply:
haha okay, that makes sense! Let me know when you've added the pointer check here and I'll review again, and then can we chat again when you are able about next TODO?